File: WdtcCore/src/main/java/org/wdt/wdtc/core/download/SpeedOfProgress.java
package org.wdt.wdtc.core.download;

import lombok.Getter;
import lombok.SneakyThrows;

import java.util.concurrent.CountDownLatch;

public class SpeedOfProgress {
  private final CountDownLatch countDown;
  @Getter
  private volatile int Spend;

  public SpeedOfProgress(int spend) {
    this.countDown = new CountDownLatch(spend);
    Spend = spend;
  }

  public synchronized void countDown() {
    Spend = Spend - 1;
    countDown.countDown();
  }


  @SneakyThrows
  public void await() {
    countDown.await();
  }

  public boolean isSpendZero() {
    return Spend == 0;
  }
}

File: WdtcCore/src/main/kotlin/org/wdt/wdtc/core/download/SpeedOfProgress.kt
package org.wdt.wdtc.core.download

import kotlinx.coroutines.CoroutineScope
import org.wdt.wdtc.core.utils.ioCoroutineScope
import java.util.concurrent.CountDownLatch
import kotlin.concurrent.Volatile

class SpeedOfProgress(@field:Volatile private var spend: Int) {

  private val countDown: CountDownLatch = CountDownLatch(spend)

  var coroutineScope: CoroutineScope = ioCoroutineScope

  @Synchronized
  fun countDown() {
    spend -= 1
    countDown.countDown()
  }

  fun await() {
    countDown.await()
  }

  val isSpendZero: Boolean
    get() = spend == 0
}
